Tests/SwiftLexicalLookupTest/Assertions.swift

```diff
@@ -0,0 +1,83 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+@_spi(Experimental) import SwiftLexicalLookup
+import SwiftParser
+import SwiftSyntax
+import XCTest
+import _SwiftSyntaxTestSupport
+
+/// Parse `source` and check if the method passed as `methodUnderTest` produces the same results as indicated in `expected`.
+///
+/// The `methodUnderTest` provides test inputs taken from the `expected` dictionary. The closure should return result produced by the tested method as an array with the same ordering.
+///
+/// - Parameters:
+///   - methodUnderTest: Closure with the tested method. Provides test argument from `expected` to the tested function. Should return method result as an array.
+///   - expected: A dictionary with parameter markers as keys and expected results as marker arrays ordered as returned by the test method.
+func assertLexicalScopeQuery(
+  source: String,
+  methodUnderTest: (SyntaxProtocol) -> ([SyntaxProtocol?]),
+  expected: [String: [String?]]
+) {
+  // Extract markers
+  let (markerDict, textWithoutMarkers) = extractMarkers(source)
+
+  // Parse the test source
+  var parser = Parser(textWithoutMarkers)
+  let sourceFileSyntax = SourceFileSyntax.parse(from: &parser)
+
+  // Iterate through the expected results
+  for (marker, expectedMarkers) in expected {
+    // Extract a test argument
+    guard let position = markerDict[marker],
+      let testArgument = sourceFileSyntax.token(at: AbsolutePosition(utf8Offset: position))
+    else {
+      XCTFail("Could not find token at location \(marker)")
+      continue
+    }
+
+    // Execute the tested method
+    let result = methodUnderTest(testArgument)
+
+    // Extract the expected results for the test argument
+    let expectedValues: [SyntaxProtocol?] = expectedMarkers.map { expectedMarker in
+      guard let expectedMarker else { return nil }
+
+      guard let expectedPosition = markerDict[expectedMarker],
+        let expectedToken = sourceFileSyntax.token(at: AbsolutePosition(utf8Offset: expectedPosition))
```

Comment:
Do we even need to extract the expected tokens? Shouldn’t it be sufficient to make `expectedValues` an `[AbsolutePosition]`? That way you don’t need to do the `firstToken` dance below.

Reply:
That's a good point, we don't really need to extract and compare tokens themselves.